Source/WebCore/css/CSSCounterStyle.cpp

```diff
@@ -27,52 +27,211 @@
+        return { };
+    unsigned valueOffset = value - firstSymbolValueForFixedSystem();
+    if (valueOffset >= symbols().size())
+        return { };
+    return symbols().at(valueOffset);
 }
-String CSSCounterStyle::counterForSystemSymbolic(int)
+
+// https://www.w3.org/TR/css-counter-styles-3/#symbolic-system
+String CSSCounterStyle::counterForSystemSymbolic(int value)
 {
-    // FIXME: implement counter for system rdar://103648354
-    return ""_s;
+    auto amountOfSymbols = symbols().size();
+    if (!amountOfSymbols || !value) {
+        ASSERT_NOT_REACHED();
+        return { };
+    }
+    unsigned symbolIndex = ((value - 1) % amountOfSymbols);
+    int frequency = static_cast<int>(std::ceil(static_cast<float>(value) / amountOfSymbols));
+
+    StringBuilder result;
+    for (int i = 0; i < frequency; ++i)
+        result.append(symbols().at(symbolIndex));
+    return result.toString();
 }
-String CSSCounterStyle::counterForSystemAlphabetic(int)
+
+// https://www.w3.org/TR/css-counter-styles-3/#alphabetic-system
+String CSSCounterStyle::counterForSystemAlphabetic(int value)
 {
-    // FIXME: implement counter for system rdar://103648354
-    return ""_s;
+    auto amountOfSymbols = symbols().size();
+    if (value < 1 || !amountOfSymbols) {
+        ASSERT_NOT_REACHED();
+        return { };
+    }
+
+    Vector<String> reversed;
+    while (value) {
+        value -= 1;
+        reversed.append(symbols().at(value % amountOfSymbols));
+        value = static_cast<int>(std::floor(value / amountOfSymbols));
+    }
+    StringBuilder result;
+    for (auto iter = reversed.rbegin(); iter != reversed.rend(); ++iter)
+        result.append(*iter);
+    return result.toString();
 }
-String CSSCounterStyle::counterForSystemNumeric(int)
+
+// https://www.w3.org/TR/css-counter-styles-3/#numeric-system
+String CSSCounterStyle::counterForSystemNumeric(int value)
 {
-    // FIXME: implement counter for system rdar://103648354
-    return ""_s;
+    auto amountOfSymbols = symbols().size();
+    if (!amountOfSymbols) {
+        ASSERT_NOT_REACHED();
+        return { };
+    }
+    if (!value)
+        return symbols().at(0);
+
+    Vector<String> reversed;
+    while (value) {
+        reversed.append(symbols().at(value % amountOfSymbols));
+        value = static_cast<int>(std::floor(value / amountOfSymbols));
+    }
+    StringBuilder result;
+    for (auto iter = reversed.rbegin(); iter != reversed.rend(); ++iter)
+        result.append(*iter);
+    return result.toString();
 }
-String CSSCounterStyle::counterForSystemAdditive(int)
+
+// https://www.w3.org/TR/css-counter-styles-3/#additive-system
+String CSSCounterStyle::counterForSystemAdditive(int value)
 {
-    // FIXME: implement counter for system rdar://103648354
-    return ""_s;
+    auto& aSymbols = additiveSymbols();
+    if (!value) {
+        for (auto& [symb, weight] : aSymbols) {
+            if (!weight)
+                return symb;
+        }
+        return { };
+    }
```

Comment:
can we avoid abbreviations? `additiveSymbols`/`symbol`/`weight`, same below

Reply:
these are the names for the getters, so can't re-use as variable. But agree that we can probably come up with other names.